### DISCUSSION TITLE
!Warning: Impact on User-Specific Caching in a SaaS Model

### DISCUSSION THREAD

**Discussion by ChoaibMouhrach** [vercel/next.js#61889] [App Router]
Hello,

I have a question after going through the documentation. I'm using a SaaS model and have a function like this:

```typescript
export const getAuthUserAction = unstable_cache(
  async () => {
    const session = await getAuthUser();

    if (!session?.user.id) {
      return session;
    }

    const user = await db.query.users.findFirst({
      where: eq(users.id, session.user.id),
    });

    if (!user) {
      return session;
    }

    session.user.email = user.email;
    session.user.image = user.image;
    session.user.name = user.name;

    return session;
  },
  ["auth"],
  {
    tags: ["auth"],
  }
);
```

I also have other actions with `revalidateTag("auth")`. Does this mean that when a user revalidates the cache, it affects the cache for all users?

If yes,

Is there a way to create unique cache keys for each user?

If no,

Doesn't this suggest that the caching system is not very useful for this type of application?

**Comment by vickonrails** [comment]
I have this exact confusion. 

Maybe someone can clarify this. 
Is the next.js `unstable_cache` doing app-wide caching or is it somehow scoped for the user?

**Reply by jepsn1** [reply]
Short answer is that it is app wide.

You can make it unique to the user by either having the user id as part of the cache key, or pass the user id to the function, as arguments are added as part of the cache key.

Now what I don't understand is whether you invalidate the cache for all users if you use revalidateTag for instance or how you would control that I have no idea.

Just saw this on reddit:

_If you want to revalidate the posts just for a specific user then use revalidateTag and add the user id as a cache tag, eg user:1 otherwise you will invalidate all user posts with the more general tag_

That makes sense I guess, so use a tag that identifies the data type combined with the user id, and then also make sure it has a unique cache key so you don't share the cache with another user accidentally.
